File: kaex/models/deployment.py
from kaex.models.resource import Resource
from kaex.models.pvc import generatePVCName


def generateContainer(app):
    container = dict()

    volumeMounts = list()
    for volume in app.volumes:
        volumeMounts.append({
            'mountPath': volume['path'],
            'name': generatePVCName(app.name, volume['path'])
        })

    return {
        'name': 'app',
        'image': f'{app.image["uri"]}:{app.image["version"]}',
        'env': app.env,
        'volumeMounts': volumeMounts
    }


def generateTemplateSpec(app):
    imagePullSecrets = list()

    if 'imagePullSecret' in app.image:
        imagePullSecrets.append({'name': app.image['imagePullSecret']})

    volumes = list()
    for volume in app.volumes:
        volumes.append({
            'name': generatePVCName(app.name, volume['path']),
            'persistentVolumeClaim': {
                'claimName': generatePVCName(app.name, volume['path'])
            }
        })

    return {
        'containers': [generateContainer(app)],
        'imagePullSecrets': imagePullSecrets,
        'volumes': volumes
    }


def generateTemplate(app):
    labels = {
        'app': app.name
    }

    return {
        'metadata': {'labels': labels},
        'spec': generateTemplateSpec(app)
    }


def generateDeploymentSpec(app):
    return {
        'replicas': app.replicas,
        'template': generateTemplate(app),
        'selector': {
            'matchLabels': {
                'app': app.name
            }
        }
    }


class Deployment(Resource):
    def __init__(self, app):
        self.apiVersion = 'apps/v1'
        self.kind = 'Deployment'

        self.metadata = {
            'name': app.name
        }

        self.spec = generateDeploymentSpec(app)

File: pkg/api/deployment.go
package api

import (
	"fmt"
	v1 "k8s.io/api/apps/v1"
	v12 "k8s.io/api/core/v1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
)

var (
	deploymentTemplate = v1.Deployment{
		TypeMeta:   metav1.TypeMeta{
			Kind:       "Deployment",
			APIVersion: "apps/v1",
		},
		ObjectMeta: metav1.ObjectMeta{},
		Spec:       v1.DeploymentSpec{
			Replicas:                nil,
			Selector:                &metav1.LabelSelector{},
			Template:                v12.PodTemplateSpec{
				ObjectMeta: metav1.ObjectMeta{
					Name:                       "",
					Annotations:                nil,
				},
				Spec:       v12.PodSpec{
					Volumes:                       nil,
				},
			},
		},
	}
)

func CreateContainers(app Application) []v12.Container {
	var envVars []v12.EnvVar
	for key, value := range app.Environment {
		envVars = append(envVars,v12.EnvVar{ Name: key, Value: value })
	}
	
	volumeMounts := make([]v12.VolumeMount, len(app.Volumes))
	for index, volume := range app.Volumes {
		for path := range volume {
			volumeMounts[index] = v12.VolumeMount{
				Name: CreatePVCName(app, path),
				MountPath: path,
			}
		}
	}

	containers := []v12.Container{{
		Name: app.Name,
		Image: fmt.Sprintf("%s:%s", app.Image, app.Version),
		Env: envVars,
		VolumeMounts: volumeMounts,
	}}
	
	return containers
}

func CreateVolumes(app Application) []v12.Volume {
	volumes := make([]v12.Volume, len(app.Volumes))
	
	for index, volume := range app.Volumes {
		for path := range volume {
			volumes[index] = v12.Volume{
				Name:         CreatePVCName(app, path),
				VolumeSource: v12.VolumeSource{
					PersistentVolumeClaim: &v12.PersistentVolumeClaimVolumeSource{
						ClaimName: CreatePVCName(app, path),
					},
				},
			}

			break
		}
	}
	
	return volumes
}

func CreateDeployment(app Application) (v1.Deployment, error) {
	deployment := deploymentTemplate
	
	deployment.ObjectMeta.Name = app.Name
	deployment.ObjectMeta.Namespace = app.Namespace
	
	if app.Replicas == 0 {
		app.Replicas = 1
	}
	deployment.Spec.Replicas = &app.Replicas

	deployment.Spec.Selector.MatchLabels = map[string]string{
		"app": app.Name,
	}
	
	if app.ImagePullSecret != "" {
		deployment.Spec.Template.Spec.ImagePullSecrets = []v12.LocalObjectReference{
			{Name: app.ImagePullSecret},
		}
	}

	deployment.Spec.Template.ObjectMeta.Labels = map[string]string{
		"app": app.Name,
	}

	deployment.Spec.Template.Spec.Volumes = CreateVolumes(app)
	deployment.Spec.Template.Spec.Containers = CreateContainers(app)

	return deployment, nil
}